### PR TITLE
[BugFix] Fixed a precision issue with one-word answers.

### LINKED ISSUE #3341: [Bug] [CI failure]: CI audio-text consistency assertion fails intermittently
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Environment: CI environment (path observed in logs: /workspace/build/buildkite/), likely a development/test container.
Full `collect_env.py` output not available — please attach.
```
</details>

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text
Not provided — reproduced around 2026-05-04
```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text
Not provided — reproduced around 2026-05-04
```
</details>

### Description
The CI logs repeatedly show the same failure: the assertion that compares the generated audio content against the reference text content frequently fails, making the e2e test flaky / failing.

For details, please check https://buildkite.com/vllm/vllm-omni/builds/8851/canvas?sid=019df433-6f85-47af-b03c-59f13e91a562&tab=output

Examples from the CI log:

```
audio content is:  London, Teotia.
text content is: London
cosine similarity text1 is: london teotia, text2 is: london
Similarity assertion failed, retrying 2/3: AssertionError('The audio content is not same as the text')
...
audio content is:  London, Paul Ting.
text content is: London
cosine similarity text1 is: london paul ting, text2 is: london
Similarity assertion failed, retrying 3/3: AssertionError('The audio content is not same as the text')
...
audio content is:  London the specialist, Ex-Sexual or London lies.
text content is: London
cosine similarity text1 is: london the specialist exsexual or london lies, text2 is: london
FAILED
```

In some runs the similarity is `1.0` (perfectly consistent), but in others the audio transcription contains extra trailing text, which causes the assertion to fail. The case is retried up to 3 times and still fails.

Throughout the test runs there is also a persistent warning:

```
RuntimeWarning: Failed to import version from _version.py: No module named 'vllm_omni._version'
This typically happens in development mode before building.
Using fallback version 'dev'.
```

Every run falls back to the `dev` version.

#### Affected file / test
- The failure occurs in `tests/e2e/online_serving/test_qwen3_omni_expansion.py::test_speaker_001[async_chunk]`
- `NIXL is available` — does not appear to be related to the failure.

### Possible root causes
- The Qwen3-Omni speaker decoding / stop-condition logic (EOS / max_tokens) may not be applied correctly, so the model keeps generating beyond `London`.
- Sampling parameters (temperature / top_p) or the random seed may not be fixed, leading to non-deterministic output.
- The similarity threshold in the test may be too strict for very short references (a single word like `London`).

### How to reproduce
Reproduces with relatively high frequency through the automated CI pipeline.

### Expected behavior
The audio content should match the reference text content, or the assertion should tolerate small / reasonable deviations. The similarity assertion logic needs to be revisited so the test is not flaky.

### Additional context
- For each failure the saved `audio data is saved: ./test_*.wav` files are available and can be used for debugging.
- The missing `vllm_omni._version` does not directly cause the failure, but it would be good to ensure `_version.py` is generated before tests run so the fallback warning goes away.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by amy-why-3459**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

Fix https://github.com/vllm-project/vllm-omni/issues/3341

When the thinker finishes, the talker needs to use an EOS token to mark that the thinker output is complete. The incorrect use of "finished" as the marker caused the request to fail to concatenate EOS, resulting in the talker outputting extra tokens.

image

## Test Plan

`pytest -sv tests/e2e/online_serving/test_qwen3_omni_expansion.py::test_one_word_prompt_001 -m "full_model" --run-level "full_model" --count=30`

## Test Result
```
=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

../../../usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: 14 warnings
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/e2e/online_serving/test_qwen3_omni_expansion.py::test_one_word_prompt_001[async_chunk-1-30]
  /usr/local/lib/python3.12/dist-packages/pydub/utils.py:14: DeprecationWarning: 'audioop' is deprecated and slated for removal in Python 3.13
    import audioop

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
--- Running Summary
================= 30 passed, 17 warnings in 1264.33s (0:21:04) =================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: tests/examples/online_serving/test_qwen3_omni.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def test_send_multimodal_request_002(omni_server) -> None:
     )
     similarity = cosine_similarity_text(audio_content.lower(), text_content.lower())
     print(f"similarity is: {similarity}")
-    assert similarity > 0.9, "The audio content is not same as the text"
+    assert similarity > 0.8, "The audio content is not same as the text"
```
Comment (hsliuustc0106) [inline tests/examples/online_serving/test_qwen3_omni.py:98]: why relax to 0.8?

Comment (amy-why-3459) [reply tests/examples/online_serving/test_qwen3_omni.py:98]: > why relax to 0.8?

After communicating with @yenuo26 , in order to eliminate the influence of whisper, we relaxed the similarity to 0.8.

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 2]
```diff
@@ -986,9 +986,11 @@ def _thinker_decode_to_talker_decode(
         start_index = meta.get("num_processed_tokens", 0)
         thinker_output_token_ids = ids.get("output", [])
         if start_index >= len(thinker_output_token_ids) - 1:
-            if meta.get("finished"):
+            # When the tokens output by the thinker are exhausted, an EOS token needs to be appended.
+            # Use the finished_flag to mark that all tokens output by thinker have been consumed.
+            if meta.get("finished_flag", False):
```
Comment (gcanlin) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py]: Maybe use `eos_emitted`, `finished_flag` was confused to me.

```suggestion
            if meta.get("eos_emitted", False):
```

Comment (hsliuustc0106) [reply vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py]: let the CI finish first

File: vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py [hunk 3]
```diff
@@ -986,9 +986,11 @@ def _thinker_decode_to_talker_decode(
...
                 return self.tts_pad_embed.to(device)
-            update_dict.setdefault("meta", {})["finished"] = True
+            update_dict.setdefault("meta", {})["finished_flag"] = True
```
Comment (gcanlin) [inline vllm_omni/model_executor/models/qwen3_omni/qwen3_omni.py]:
```suggestion
            update_dict.setdefault("meta", {})["eos_emitted"] = True
```

**Comment by oglok** [comment]
I closed the temperature fix, I think this is a better approach. However the PR seems a bit unstructured.

- All the commented code and tests for Bagel are unrelated to this issue
- If this fix works, there is no need to increase the amount of retries from 3 to 10.
- It'd be good to explain in the PR description that there is a collision between the stage_input_processors pipeline meta keys and the model layer.

**Comment by amy-why-3459** [comment]
> I closed the temperature fix, I think this is a better approach. However the PR seems a bit unstructured.
> 
> * All the commented code and tests for Bagel are unrelated to this issue
> * If this fix works, there is no need to increase the amount of retries from 3 to 10.
> * It'd be good to explain in the PR description that there is a collision between the stage_input_processors pipeline meta keys and the model layer.

Thank you very much for your suggestion. I will add a description of the changes in the PR.

**Review by hsliuustc0106 — COMMENTED** [review]
**BLOCKING:**

- **Scope** — This PR mixes the core bugfix with unrelated changes: commented Bagel tests, test threshold relaxations, and retry increases. The collision fix ( → ) and test evidence are solid, but please address the structural issues flagged by oglok:

1. Remove commented Bagel code and test changes not related to this issue
2. Explain why  avoids the  pipeline meta collision in the PR description
3. If the EOS fix is correct, the retry increase from 3 to 10 should not be needed
4. Justify the similarity threshold changes from 0.9 to 0.8

The core fix (avoiding key collision) looks reasonable, but the PR needs restructuring before merge.

**Review by hsliuustc0106 — COMMENTED** [review]
BLOCKING:

Scope - This PR mixes the core bugfix with unrelated changes: commented Bagel tests, test threshold relaxations, and retry increases. The key collision fix looks reasonable, but please address the structural issues flagged by oglok:

1. Remove commented Bagel code and test changes not related to this issue
2. Explain why the meta key change avoids collision with stage_input_processors pipeline in the PR description
3. If the EOS fix is correct, the retry increase from 3 to 10 should not be needed
4. Justify the similarity threshold changes from 0.9 to 0.8

**Comment by oglok** [comment]
It's already merged but the suggestions were not addressed.

**Comment by amy-why-3459** [comment]
> It's already merged but the suggestions were not addressed.

1. Remove commented Bagel code and test changes not related to this issue

    **The test cases for bagel were removed because they were not ready and had been commented out.**

2. Explain why the meta key change avoids collision with stage_input_processors pipeline in the PR description

    **Description has been added to the PR**

3. If the EOS fix is correct, the retry increase from 3 to 10 should not be needed

    **The number of retries was changed from 3 to 10 to eliminate the effect of whisper.**

4. Justify the similarity threshold changes from 0.9 to 0.8
     **The threshold was changed from 0.9 to 0.8, also to eliminate the effect of whisper.**

bb914ef919ad0be5de4c3ad285ed3b2e
   

The third and fourth points concern the modification of the test cases, which has been agreed upon with the test case owner @yenuo26 .

**Comment by oglok** [comment]
> > It's already merged but the suggestions were not addressed.
> 
> 1. Remove commented Bagel code and test changes not related to this issue
>    **The test cases for bagel were removed because they were not ready and had been commented out.**

These test cases were not related to the purpose of the fix, now, another developer had to write a PR to clean it up. See: https://github.com/vllm-project/vllm-omni/pull/3407

> 2. Explain why the meta key change avoids collision with stage_input_processors pipeline in the PR description
>    **Description has been added to the PR**

> 3. If the EOS fix is correct, the retry increase from 3 to 10 should not be needed
>    **The number of retries was changed from 3 to 10 to eliminate the effect of whisper.**

Again, if the fix works, we should not need 10 retries. 3 should be enough.

> 4. Justify the similarity threshold changes from 0.9 to 0.8
>    **The threshold was changed from 0.9 to 0.8, also to eliminate the effect of whisper.**
> 
> bb914ef919ad0be5de4c3ad285ed3b2e
> The third and fourth points concern the modification of the test cases, which has been agreed upon with the test case owner @yenuo26 .

**Comment by amy-why-3459** [comment]
> > > It's already merged but the suggestions were not addressed.
> > 
> > 
> > 
> > 1. Remove commented Bagel code and test changes not related to this issue
> >    **The test cases for bagel were removed because they were not ready and had been commented out.**
> 
> These test cases were not related to the purpose of the fix, now, another developer had to write a PR to clean it up. See: #3407
> 
> > 2. Explain why the meta key change avoids collision with stage_input_processors pipeline in the PR description
> >    **Description has been added to the PR**
> 
> > 3. If the EOS fix is correct, the retry increase from 3 to 10 should not be needed
> >    **The number of retries was changed from 3 to 10 to eliminate the effect of whisper.**
> 
> Again, if the fix works, we should not need 10 retries. 3 should be enough.
> 
> > 4. Justify the similarity threshold changes from 0.9 to 0.8
> >    **The threshold was changed from 0.9 to 0.8, also to eliminate the effect of whisper.**
> > 
> > bb914ef919ad0be5de4c3ad285ed3b2e
> > The third and fourth points concern the modification of the test cases, which has been agreed upon with the test case owner @yenuo26 .

Thank you very much for your suggestion. However, this test case was originally supposed to be deleted. I simply commented it out and am waiting for @fake0fan to completely remove it. You can check out the comments on this PR: https://github.com/vllm-project/vllm-omni/pull/2396.
